pkg/storage/minio/saver.go

```diff
@@ -1,34 +1,35 @@
 package minio
 
 import (
-	"bytes"
 	"context"
-	"io"
 
 	"github.com/gomods/athens/pkg/errors"
 	"github.com/gomods/athens/pkg/observ"
+	"github.com/gomods/athens/pkg/storage"
+	moduploader "github.com/gomods/athens/pkg/storage/module"
 	minio "github.com/minio/minio-go"
 )
 
-func (s *storageImpl) Save(ctx context.Context, module, vsn string, mod []byte, zip io.Reader, info []byte) error {
-	const op errors.Op = "storage.minio.Save"
+const mb = 1000000
+
+func (s *storageImpl) Save(ctx context.Context, module, version string, mod []byte, zip storage.Zip, info []byte) error {
+	const op errors.Op = "minio.Save"
 	ctx, span := observ.StartSpan(ctx, op.String())
 	defer span.End()
-	dir := s.versionLocation(module, vsn)
-	modFileName := dir + "/" + "go.mod"
-	zipFileName := dir + "/" + "source.zip"
-	infoFileName := dir + "/" + vsn + ".info"
-	_, err := s.minioClient.PutObject(s.bucketName, modFileName, bytes.NewReader(mod), int64(len(mod)), minio.PutObjectOptions{})
-	if err != nil {
-		return errors.E(op, err)
-	}
-	_, err = s.minioClient.PutObject(s.bucketName, zipFileName, zip, -1, minio.PutObjectOptions{})
-	if err != nil {
-		return errors.E(op, err)
-	}
-	_, err = s.minioClient.PutObject(s.bucketName, infoFileName, bytes.NewReader(info), int64(len(info)), minio.PutObjectOptions{})
+
+	zipStream := moduploader.Stream{Stream: zip.Reader, Size: zip.Size}
+	err := moduploader.Upload(ctx, module, version, moduploader.NewStreamFromBytes(info), moduploader.NewStreamFromBytes(mod), zipStream, s.upload, s.timeout)
 	if err != nil {
-		return errors.E(op, err)
+		return errors.E(op, err, errors.M(module), errors.V(version))
 	}
 	return nil
 }
+
+func (s *storageImpl) upload(ctx context.Context, path, contentType string, stream moduploader.Stream) error {
+	const op errors.Op = "minio.upload"
+	if stream.Size > 600*mb {
+		stream.Size = -1
```

Comment:
similar to my question below, why can't we just put -1 for all the things? This would guarantee the same behavior for all modules